Themes: Theme carousel

0nko opened the pull request:
Fixes #10234. The PR adds the theme picker screen with the theme carousel.

<image src="https://github.com/woocommerce/woocommerce-android/assets/1522856/51fdefc9-8178-4513-901c-ca25cba6dd6d" width="300" />

**To test:**
1. Start the app and trigger the store creation
2. Go through the profiler steps
3. After the feature profiler screen, notice the next screen is the theme picker screen (only when feature flag is enabled)
4. Notice a couple of demo site screenshots are loaded (some are not working, they're just placeholders)

0nko committed:
Add a new feature flag

0nko committed:
Add a new theme picker screen

0nko committed:
Add navigation to/from theme picker

0nko committed:
Update the navigation in the feature profiler screen

0nko committed:
Add string resources

0nko committed:
Add carousel UI and data

hafizrahman commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/storecreation/themes/ThemePickerViewModel.kt:
```diff
@@ -0,0 +1,70 @@
+                    CarouselItem.Theme(screenshotUrl = getThemeUrl(themeName = "tazza")),
+                    CarouselItem.Theme(screenshotUrl = getThemeUrl(themeName = "amulet")),
+                    CarouselItem.Theme(screenshotUrl = getThemeUrl(themeName = "zaino")),
+                    CarouselItem.Theme(screenshotUrl = getThemeUrl(themeName = "thriving-artist")),
```
For this one the correct demo URL is:

https://thrivingartistdemo.wpcomstaging.com/

hichamboushaba replied:
The demo links will be fetched from the API, I think it's OK to keep those until we connect this with the networking layer.

hafizrahman commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/storecreation/themes/ThemePickerViewModel.kt:
```diff
@@ -0,0 +1,70 @@
+                    CarouselItem.Theme(screenshotUrl = getThemeUrl(themeName = "amulet")),
+                    CarouselItem.Theme(screenshotUrl = getThemeUrl(themeName = "zaino")),
+                    CarouselItem.Theme(screenshotUrl = getThemeUrl(themeName = "thriving-artist")),
+                    CarouselItem.Theme(screenshotUrl = getThemeUrl(themeName = "attar")),
```
For this one the correct demo URL is:

https://attardemo.wordpress.com/

hafizrahman reviewed (changes requested):
On landscape the screenshots are not very visible, so perhaps we need to make it bigger:

![Screenshot_20231127_150153](https://github.com/woocommerce/woocommerce-android/assets/266376/a5b00450-bc31-46f1-9209-400e6249717c)

hafizrahman commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/storecreation/themes/ThemePickerScreen.kt:
```diff
@@ -0,0 +1,158 @@
+        ) {
+            Text(
+                text = title,
+                style = MaterialTheme.typography.subtitle1,
```
I think "Looking for more" is bold in the design:

<img src="https://github.com/woocommerce/woocommerce-android/assets/266376/faeb50ab-f7d6-4309-bad9-7be534d7802f" width="50%"/>

hafizrahman reviewed (commented):
@0nko Nice work, and I truly enjoy how simple the carousel code is with the help of Jetpack Compose. It works quite nicely with the scrolling too.

I left some comment & questions. For the theme demo URLs specifically it's perhaps easier to hardcode the URLs since turns out they don't always use wpcomstaging.com site.

I also see there's no "Skip" button on the toolbar but not sure if that's meant for a separate PR.

Also curious about the bot error message here: https://github.com/woocommerce/woocommerce-android/pull/10252#issuecomment-1825968951

Seems like something recently added, and I'm not sure what's the policy for that now (must handle now, handle later in separate PR, something else?)

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/storecreation/themes/ThemePickerFragment.kt:
```diff
@@ -0,0 +1,57 @@
+package com.woocommerce.android.ui.login.storecreation.themes
```
@0nko, I want to add a `Repository` for the networking, and I just wanted to confirm the package name we should be using.
Since we'll have an entry to this from Settings, I think we shouldn't put it in `storecreation`, WDYT? 

There are some slight differences between the version we'll show as a profiler step and the version we'll show in the settings, but I think we should still aim to reuse most of the screen, WDYT?

0nko replied:
That makes sense. Do you want to put it completely to a separate package, like `com.woocommerce.android.ui.themes`?

0nko replied:
Just checked your repository PR and saw it. I'll move my code there.

hichamboushaba replied:
Sorry, I missed the ping for your reply, thanks for taking care of this.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/storecreation/themes/ThemePickerScreen.kt:
```diff
@@ -0,0 +1,158 @@
+        AsyncImage(
+            model = ImageRequest.Builder(LocalContext.current)
+                .data(screenshotUrl)
+                .diskCachePolicy(DISABLED)
```
Curious: why disable caching here?

And do we need to have some loading and error placeholders? There is an ongoing discussion about failures here pe5sF9-2f9-p2#comment-3135 that might be of interest.

hafizrahman replied:
I was going to mention the above, too. I think perhaps failure handling can be handled in a separate PR, but want to know about disabling cache as well.

0nko replied:
I forgot to get rid of this. I tried to force-reload the image when I was implementing the carousel.

0nko committed:
Make the message title bold

0nko committed:
Move the themes classes to a different package

0nko committed:
Rename string resource names

0nko committed:
Add the skip button

0nko committed:
Disable theme image load redirect and handle errors

0nko committed:
Fix ktlint issue

0nko committed:
Make the theme picker vertically scrollable

0nko commented:
@hafizrahman @hichamboushaba Thank you both for the review. 

I struggled a bit to make the "generating" case behave properly. It now works pretty well. The suggested text message is shown for errors (you can try turning on the flight mode). Also, since redirects are treated as errors now, when you flick around the carousel for a few seconds, the generated screenshot is reloaded automatically (I added some random URLs to demonstrate it).

> I also see there's no "Skip" button on the toolbar but not sure if that's meant for a separate PR.

There was no Skip button in the Android designs but I suppose it was omitted by mistake.

> Also curious about the bot error message here: https://github.com/woocommerce/woocommerce-android/pull/10252#issuecomment-1825968951

Yeah, I added a label to ignore the error as I plan to add tests when the functionality is complete.

I believe I addressed all comments -- ready for another round.

hafizrahman reviewed (approved):
👍🏼  LGTM, thanks for tackling the feedbacks!
